codeU5.py

```diff
@@ -0,0 +1,127 @@
+
+class Alphabet:
+    def __init__(self,word_list):
+        
+        '''Args:
+            word_list: non-empty list of unique words from the given language
+        '''
+        
+        self.words = word_list
+        self.partial_order = [] # a list of partial order lists [char1,char2], char1 < char2
+        self.all_chars = set() # unordered set of all characters
+        self.sorted_alphabet = [] # the list of characters in the alphabetical order
+        for word in self.words:
+            self.all_chars.update(set(word))
+        self.neigh_dict = defaultdict(list) # dictionary of char: [list of chars going after the given char] 
+        
+    def find_common_prefix(self, word1, word2):
+        
+        if len(word1) < len(word2):
+            word_shorter = word1
+            word_longer = word2
+        else:
+            word_longer = word1
+            word_shorter = word2
+            
+        for i, c in enumerate(word_shorter):
+            if c != word_longer[i]:
+                return word_shorter[:i]
+    
+    def find_partial_order(self, word1, word2):
+        
+        '''Returns:
+            partial order [char_prev,char_next]
+            inferred from two words word1 and word2
+        '''
+        
+        prefix = self.find_common_prefix(word1, word2)
+        char_first = word1[len(prefix)]
+        char_next = word2[len(prefix)]
+        return [char_first, char_next]
+    
+    def find_all_partial_orders(self):
+        
+        '''Returns:
+            all partial orders inferred from
+            each subsequent pair of words
+        '''
+        
+        for i in range(len(self.words) - 1):
+            # if one word is contained in the other, then we can not infer the partial order
+            if self.words[i] in self.words[i+1] or self.words[i+1] in self.words[i]:
+                continue    
+            [char_prev, char_next] = self.find_partial_order(self.words[i], self.words[i+1])
+            self.partial_order.append([char_prev, char_next])
+        return self.partial_order 
+    
+    def find_neighbours(self):
+        
+        '''Creates a dictionary 
+        with key as a char and value being as a list of chars
+        that go next in the alphabetical order
+        '''
+        
+        for [char_prev, char_next] in self.partial_order:
+            if char_prev not in self.neigh_dict:
+                self.neigh_dict[char_prev] = [char_next]
+            else:
+                self.neigh_dict[char_prev]+= [char_next]
+    
+    def topological_sort(self):
+        
+        '''The implementation of topological sort
+        based on DFS, check https://en.wikipedia.org/wiki/Topological_sorting
+        Returns:
+            sorted alphabet 
+        '''
+        
+        self.find_all_partial_orders()
+        self.find_neighbours()
+        self.temp_marked = set()
+        self.unmarked = self.all_chars
+        unmarked = self.unmarked.copy()
+        for node in unmarked:

```

Comment:
It looks strange to `self.unmarked.copy()` here only to check for memberhsip right in `visit_node`. Is this so that the `for` loop here could work over an unchanging `unmarked`?

Alternatively, could consider:

```
while self.unmarked:
  node = self.unmarked.pop()
  self.visit_node(node)
```